lib/runner/ssh-tunnel-agent.js

```diff
@@ -0,0 +1,221 @@
+var SshClient = require('ssh2'),
+    _ = require('lodash'),
+    ProxyConfigList = require('postman-collection').ProxyConfigList,
+    net = require('net'),
+    fs = require('fs'),
+    EventEmitter = require('events').EventEmitter,
+
+    HIGHEST_POSSIBLE_PORT = 65535,
+    LOCALHOST = '127.0.0.1',
+    DEFAULT_LOCAL_PORT = 0,
+
+    SshTunnelAgent;
+
+/**
+ * Tunnel Agent is responsible for managining and starting new SSH tunnels, it can establish tunnel
+ * between mutiple SSH hosts and can can route the request to correct host by matching the against
+ * url.
+ *
+ * @constructor
+ * @extends Events
+ * @param {ProxyConfigList} proxyConfigList
+ */
+SshTunnelAgent = function SshTunnelAgent (proxyConfigList) {
+    this.proxyConfigList = _.assign({}, proxyConfigList);
+    this.mapping = [];
+    this.readyTunnelCount = 0;
+    this.failedTunnelCount = 0;
+    this.reporter = new EventEmitter();
+};
+
+_.assign(SshTunnelAgent.prototype, {
+    isReady: function (status) {
+        status ? this.readyTunnelCount++ : this.failedTunnelCount++;
+        if (this.readyTunnelCount + this.failedTunnelCount === this.proxyConfigList.count()) {
+            this.reporter.emit('ready', this.readyTunnelCount);
+        }
+    },
+
+    /**
+     * Given a sshconfig creates a single new tunnel between postman-runtime and ssh host.
+     *
+     * @param {SshConfig} sshConfig config related to ssh connection establishment and authentication
+     * @return {SshClient}
+     */
+    createTunnel: function (sshConfig) {
+        var sshc = new SshClient(),
+            self = this;
+
+        sshc.connect({
+            host: sshConfig.auth.host,
+            port: sshConfig.auth.port,
+            username: sshConfig.auth.username,
+            // eslint-disable-next-line security/detect-non-literal-fs-filename
+            privateKey: fs.readFileSync(sshConfig.auth.privateKeyPath)
+        });
+
+        sshc.on('error', function(err) {
+            var error = new Error('error establishing tunnel for host: ' + sshConfig.auth.host + err.message);
+
+            self.reporter.emit('tunnelError', error);
+            self.isReady(false);
+        });
+
+        sshc.on('ready', function () {
+            var update = {config: sshConfig, client: sshc};
+
+            !_.includes(self.mapping, update) && self.mapping.push(update);
```

Comment:
It's odd that an array-valued property is called `mapping` and not `list`.

Reply:
it's a list of mappings. can't call it plain list, routing list maybe?